File: packages/metadata_core/lib/src/photo_meta.dart
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================
import 'field.dart';

bool orientationSwapsAxes(int? orientation) =>
    orientation != null && orientation >= 5 && orientation <= 8;

class PhotoMeta {
  const PhotoMeta({
    required this.datetimeOriginal,
    required this.gpsLatitude,
    required this.gpsLongitude,
    required this.orientation,
    required this.make,
    required this.model,
    required this.exposureTime,
    required this.widthRaw,
    required this.heightRaw,
    required this.fileSizeBytes,
    required this.mime,
  });

  final MetaField<String> datetimeOriginal;
  final MetaField<double> gpsLatitude;
  final MetaField<double> gpsLongitude;
  final MetaField<int> orientation;
  final MetaField<String> make;
  final MetaField<String> model;
  final MetaField<String> exposureTime;
  final MetaField<int> widthRaw;
  final MetaField<int> heightRaw;
  final MetaField<int> fileSizeBytes;
  final MetaField<String> mime;

  int? get width {
    final int? rawWidth = widthRaw.value;
    final int? rawHeight = heightRaw.value;
    if (rawWidth == null || rawHeight == null) {
      return null;
    }
    return orientationSwapsAxes(orientation.value) ? rawHeight : rawWidth;
  }

  int? get height {
    final int? rawWidth = widthRaw.value;
    final int? rawHeight = heightRaw.value;
    if (rawWidth == null || rawHeight == null) {
      return null;
    }
    return orientationSwapsAxes(orientation.value) ? rawWidth : rawHeight;
  }
}

File: native/vision-core/src/main/kotlin/com/visionengine/core/PhotoMeta.kt
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================
package com.visionengine.core

fun orientationSwapsAxes(orientation: Long?): Boolean =
    orientation != null && orientation in 5L..8L

data class PhotoMeta(
    val datetimeOriginal: MetaField<String>,
    val gpsLatitude: MetaField<Double>,
    val gpsLongitude: MetaField<Double>,
    val orientation: MetaField<Long>,
    val make: MetaField<String>,
    val model: MetaField<String>,
    val exposureTime: MetaField<String>,
    val widthRaw: MetaField<Long>,
    val heightRaw: MetaField<Long>,
    val fileSizeBytes: MetaField<Long>,
    val mime: MetaField<String>,
) {
    val width: Long?
        get() {
            val rawWidth = widthRaw.value
            val rawHeight = heightRaw.value
            if (rawWidth == null || rawHeight == null) {
                return null
            }
            return if (orientationSwapsAxes(orientation.value)) rawHeight else rawWidth
        }

    val height: Long?
        get() {
            val rawWidth = widthRaw.value
            val rawHeight = heightRaw.value
            if (rawWidth == null || rawHeight == null) {
                return null
            }
            return if (orientationSwapsAxes(orientation.value)) rawWidth else rawHeight
        }
}
